### PR TITLE
Add service block to formula api

### LINKED ISSUE #14925: API install does not replace `Dir.home` when pouring bottles
### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  php-uuid@8.1

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  jsonschema
  php@7.4
  php@8.0

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  php-raphf@8.1

Warning: You have uncommitted modifications to Homebrew/homebrew-cask.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-cask && git stash -u && git clean -d -f

Uncommitted files:
   M Casks/virtualbox.rb
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.0.5-103-gbde256a
ORIGIN: https://github.com/Homebrew/brew
HEAD: bde256a325f60463f14c4d19cd40e0be99befa53
Last commit: 7 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a3a6d385a270d02e75c180949805b12c87331612
Core tap last commit: 7 hours ago
Core tap branch: master
Core tap JSON: 08 Mar 13:31 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DEVELOPER: set
HOMEBREW_EDITOR: nano
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_ENV_HINTS: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2 => /Applications/Xcode_14.2.app/Contents/Developer
Rosetta 2: false
```

### What were you trying to do (and why)?

Install any formula that uses a working_dir of `Dir.home` in the service block.

### What happened (include all command output)?

The poured services define a home directory that only applies to the CI runners
```ini
WorkingDirectory=/Users/brew/actions-runner/_work/homebrew-core/homebrew-core/bottles/home
```
```plist
        <key>WorkingDirectory</key>
        <string>/Users/brew/actions-runner/_work/homebrew-core/homebrew-core/bottles/home</string>
```

### What did you expect to happen?

For that value to be replaced by the user's home directory

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install gitlab-runner
cat $(brew --prefix gitlab-runner)/homebrew.mxcl.gitlab-runner.plist
cat $(brew --prefix gitlab-runner)/homebrew.gitlab-runner.service
```

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This fixes #14925.

The idea here is to include the service block info in the formula API JSON so that it can be loaded from the API. This will allow us to more consistently build service files and will mean that `brew services` doesn't have to be updated to work with the new API.

The first thing that was added was the ability to serialize and deserialize service blocks. We needed the ability to replace all local paths with placeholders when building the API JSON and then replace those placeholders with local paths when loading from the API.

The second thing was adding this service block JSON to the API and loading it from the API.

The other commits are just tests.

### CODE REVIEW

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [inline Library/Homebrew/formula.rb:1046]: Was there any special reason why this wasn't memoized before?

File: Library/Homebrew/formulary.rb
Comment (apainintheneck) [inline Library/Homebrew/formulary.rb:274]: `Service#run` accepts either one param or kwargs so needs to be handled separately.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [inline Library/Homebrew/service.rb:50]: Open to other ideas of how to handle this. Couldn't think of anything better.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [inline Library/Homebrew/service.rb:326]: This is only used with `Symbols` in `homebrew/core`.

```
/u/l/H/L/T/h/homebrew-core (master|✔) $ git grep environment_variables
Formula/apache-archiva.rb:    environment_variables ARCHIVA_BASE: var/"archiva"
Formula/appium.rb:    environment_variables PATH: std_service_path_env
Formula/clipper.rb:    environment_variables LANG: "en_US.UTF-8"
Formula/couchdb-lucene.rb:    environment_variables HOME: "~"
Formula/davmail.rb:    environment_variables PATH: std_service_path_env
Formula/docker-machine.rb:    environment_variables PATH: std_service_path_env
Formula/docuum.rb:    environment_variables PATH: std_service_path_env
Formula/dovecot.rb:    environment_variables PATH: std_service_path_env
Formula/ejabberd.rb:    environment_variables HOME: var/"lib/ejabberd"
Formula/etcd.rb:    environment_variables ETCD_UNSUPPORTED_ARCH: "arm64" if Hardware::CPU.arm?
Formula/gitlab-runner.rb:    environment_variables PATH: std_service_path_env
Formula/hbase.rb:    environment_variables HBASE_HOME:              opt_libexec,
Formula/httpd.rb:    environment_variables PATH: std_service_path_env
Formula/influxdb.rb:    environment_variables INFLUXD_CONFIG_PATH: etc/"influxdb2/config.yml"
Formula/isync.rb:    environment_variables PATH: std_service_path_env
Formula/jack.rb:    environment_variables PATH: "/usr/bin:/bin:/usr/sbin:/sbin:#{HOMEBREW_PREFIX}/bin"
Formula/kettle.rb:    environment_variables KETTLE_HOME: etc/"kettle"
Formula/launch_socket_server.rb:    environment_variables LAUNCH_PROGRAM_TCP_ADDRESS: "127.0.0.1:8080"
Formula/libvirt.rb:    environment_variables PATH: HOMEBREW_PREFIX/"bin"
Formula/nzbget.rb:    environment_variables PATH: "#{HOMEBREW_PREFIX}/bin:/usr/bin:/bin:/usr/sbin:/sbin"
Formula/offlineimap.rb:    environment_variables PATH: std_service_path_env
Formula/payara.rb:    environment_variables GLASSFISH_HOME: opt_libexec/"glassfish"
Formula/podman.rb:    environment_variables PATH: std_service_path_env
Formula/postgresql@15.rb:    environment_variables LC_ALL: "C"
Formula/rabbitmq.rb:    environment_variables PATH:          "#{HOMEBREW_PREFIX}/sbin:/usr/sbin:/usr/bin:/bin:#{HOMEBREW_PREFIX}/bin",
Formula/runit.rb:    environment_variables PATH: "/usr/bin:/bin:/usr/sbin:/sbin:#{opt_bin}"
Formula/wildfly-as.rb:    environment_variables JBOSS_HOME: opt_libexec, WILDFLY_HOME: opt_libexec
Formula/znapzend.rb:    environment_variables PATH: std_service_path_env
Formula/zookeeper.rb:    environment_variables SERVER_JVMFLAGS: "-Dapple.awt.UIElement=true"
Formula/zrepl.rb:    environment_variables PATH: std_service_path_env
```

File: Library/Homebrew/service.rb
Comment (apainintheneck) [inline Library/Homebrew/service.rb:533]: Could use a better comment.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [inline Library/Homebrew/service.rb:513]: Could use a better comment.

**Comment by apainintheneck** [comment]
Here's my testing script for checking that the service files were being generated correctly and that the serialization was consistent.

- Script: https://gist.github.com/apainintheneck/6ed138ac7214788cd20bb7da82f7aa36

File: Library/Homebrew/formulary.rb
Comment (SMillerDev) [reply Library/Homebrew/formulary.rb:274]: How about keep_alive?

Comment (apainintheneck) [reply Library/Homebrew/formulary.rb:274]: https://github.com/Homebrew/brew/blob/8f6276ed5f1c7b8293128a481b175611cb2980c8/Library/Homebrew/service.rb#L121-L141

That accepts a string or a hash but always gets stored internally as a hash. It's not a problem if you just pass that hash to it then. It'd be trickier if it accepted kwargs.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:1046]: Probably not.

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:50]: Seems fine to me!

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:540]: `from_hash` feels more typical naming here

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:517]: Can this get a type signature?

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:540]: Can this get a type signature?

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:521]: Feels like having a shared method to do this in `Utils` or something that can be used elsewhere will help us avoid forgetting to add more of these.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! Some style thoughts.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:540]: I guess I'd expect `from_hash` to create a service object which is not what's happening here. I'm not really happy with the naming here either though.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:521]: How would that method work? The way I see it we don't always apply or need to apply the same substitutions everywhere. We do have the `keg_relocate` code but something like that seems heavy handed for this specific case.

CC: @Rylan12 since you mentioned something similar in another thread

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:517]: We could always slap on something generic but I'm not sure how useful a specific type signature would be here. The only other option would be refactoring this into smaller functions.

@dduugg How would you approach adding typing to a recursive function like this?

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:521]: We could also do the substitutions right before or immediately after writing the `cask.json` or `formula.json` file if we wanted to have something super generic and only do the substitutions in one place.

File: Library/Homebrew/service.rb
Comment (dduugg) [reply Library/Homebrew/service.rb:517]: If I'm reading the method documentation correctly, the best you can probably do is:
```ruby
sig { params(elem: T.untyped).returns(T::Hash[String, T.untyped]) }
```
Trying to type an undefined shape generally isn't worth the effort. `JSON.dump` is probably analogous here, and sorbet just leaves it completely [untyped](https://github.com/sorbet/sorbet/blob/60fb704/rbi/stdlib/json.rbi#L130-L138).

File: Library/Homebrew/service.rb
Comment (Rylan12) [reply Library/Homebrew/service.rb:521]: I think one option is to have a method that just does all of the replacements we deem necessary (essentially `Cask::Cask.to_h_string_gsubs` before it was removed):

```ruby
def replace_strings(string)
  string.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
        .gsub(HOME, HOME_PLACEHOLDER)
```

And then we can either selectively call this where it's needed or just recursively call this on the hash generated by `Formula#to_hash` before returning

Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:521]: > just recursively call this on the hash generated by `Formula#to_hash` before returning

Something like this makes most sense to me.

I don't think it makes sense to scatter the replacements all over the codebase when they are only needed when reading/writing hashes.

CC @issyl0 as this affects https://github.com/Homebrew/brew/pull/15004 and possibly other PRs you have open

File: Library/Homebrew/service.rb
Comment (dduugg) [reply Library/Homebrew/service.rb:517]: Oh, another pattern I've seen involves typing the input, something like:
```ruby
sig { params(elem: T::Hash[Symbol, T.untyped]).returns(T::Hash[String, T.untyped]) }
```
and then have `serialize` call a private `_serialize` method that handles the recursive calls. (Ofc, if you know more about the shape of the input/output you could flesh out the type annotations accordingly.)

**Comment by apainintheneck** [comment]
Thanks for the feedback. I think that I'll rewrite the recursive (de)serialization methods to get rid of the recursion. It'll allow me to add better type annotations and I don't think the special handling around symbols will be necessary anymore which will be another bonus. It will be a little more verbose but that's probably fine.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:521]: > Yeah, the only reason I can see for not doing that is that there is at least one place we decided we didn't want it to be replaced in casks (#14595)

Yep, this is the exact reason I'm worried about applying it more generally. Also, if we know the substitutions are not needed for most of the formula hash we're just doing extra work for no reason.

Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:521]: Yeh, I forgot this method was changed in https://github.com/Homebrew/brew/pull/14615 to use monkeypatches to avoid the need for as many explicit substitutions.

I think it'd be nicer to use this approach instead wherever possible.

Comment (apainintheneck) [reply Library/Homebrew/service.rb:521]: That could work for the hash generation but we'd still need to substitute things afterwards when deserializing. I think that the generating hash code for casks can probably just be converted into a generic mixin.

**Comment by Bo98** [comment]
Ping me when this is near final and I'll review this.

**Comment by apainintheneck** [comment]
I changed the serialization methods to not be recursive to make it easier to enable typing. I also changed the paths substitutions when serializing to use the same method as we already use with casks. This involves monkey patching the `$HOME` and `$HOMEBREW_PREFIX` values when generating the hash. Beyond that I updated the script from before.

- [Script](https://gist.github.com/apainintheneck/6ed138ac7214788cd20bb7da82f7aa36)

The generated service files are still the same but the serialization is different. That is because I'm not converting the `Pathname` objects to `String`s anymore because that happens naturally when converting it to JSON later on.

You can see the different serialization results [here](https://gist.github.com/apainintheneck/8b441c2a63d41dd88f68ddd5d41e36fb).

I still need to do more testing around the `generate-*-api` commands though to make sure they're still working correctly. Also, will there be any problems with monkey patching those common paths when generating the formula api JSON? I can't think of anything but I might have missed something.

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/38146893c3966d0ad1a359f866df4dbf49b9e176

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/bc85857e1341eef2e105cbacd3e5f1e4eaaf47a0

**Comment by apainintheneck** [comment]
@Bo98 This is ready for final review.

File: Library/Homebrew/test/cask/cask_spec.rb
Comment (apainintheneck) [inline Library/Homebrew/test/cask/cask_spec.rb:341]: This comment should get updated. It's more that you can't have multiple tests for the same class.

Comment (apainintheneck) [reply Library/Homebrew/test/cask/cask_spec.rb:341]: Maybe I should add it to the `APIHashable` file instead?

**Review by SMillerDev — APPROVED** [review]

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:49]:
```suggestion
      @run_params ||= command 
      @run_params ||= { macos: macos, linux: linux }.compact
```

feels a little bit more typical/idiomatic/easy to follow

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:585]: It's more that `false` can be a valid value here so `nil` is the only value we really know is invalid.

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:25]: That was updated in another PR. Should be fine once this is merged into master.

File: Library/Homebrew/cask/cask_loader.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask_loader.rb:334]: Doesn't matter to me. Do you feel strongly either way?

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:562]: I prefer to have the whole case statement at the same indentation but the use of the method method is a good point.

File: Library/Homebrew/test/cask/cask_spec.rb
Comment (apainintheneck) [reply Library/Homebrew/test/cask/cask_spec.rb:341]: Sorry, this was more a note to self. 😁

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb:521]: I prefer it like it is.

File: Library/Homebrew/cask/cask_loader.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask_loader.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

File: Library/Homebrew/service.rb
Comment (apainintheneck) [reply Library/Homebrew/service.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

File: Library/Homebrew/test/cask/cask_spec.rb
Comment (apainintheneck) [reply Library/Homebrew/test/cask/cask_spec.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/6ccf21c2c9a563e95476d31d12c1c04eedc3deed

File: Library/Homebrew/cask/cask_loader.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask_loader.rb]: Addressed in https://github.com/Homebrew/brew/pull/15007/commits/6ccf21c2c9a563e95476d31d12c1c04eedc3deed

**Comment by apainintheneck** [comment]
I'm now happy with this as it is right now. I'll wait another day for any final reviews and then merge it in sometime tomorrow.

As a sanity check I created another script that's a sort of `brew readable` but for the generated API. Might be useful for other PRs too.
- [Readall Script](https://gist.github.com/apainintheneck/ce23cf2ca96f0d5ff988f9ab16a7f4aa)

File: Library/Homebrew/service.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:586]: All good!

**Review by MikeMcQuaid — APPROVED** [review]
Looks great, thanks again @apainintheneck!

**Comment by MikeMcQuaid** [comment]
> As a sanity check I created another script that's a sort of `brew readall` but for the generated API. Might be useful for other PRs too.

@apainintheneck 💯% yes! Would be nice to get this as a CI check in Homebrew/brew (not in/blocking this PR)? How long does it take to run for you?

**Review by carlocab — APPROVED** [review]

**Pull request merged**
